### PR TITLE
feat: support key bindings in textinput like textarea

### LINKED ISSUE #269: textinput: doesn't support customizable key bindings
As discussed on the public slack. I'll send a PR.

### PR DESCRIPTION
**Pull request by knz**
Fixes #269.

cc @muesli @maaslalani

### CODE REVIEW

**Comment by maaslalani** [comment]
Thanks @knz, this looks great! Testing now

File: textinput/textinput.go [hunk 1]
```diff
@@ -37,6 +38,41 @@ const (
...
+	DeleteBeforeCursor      key.Binding
+	DeleteCharacterBackward key.Binding
+	DeleteCharacterForward  key.Binding
+	LineStart               key.Binding
+	LineEnd                 key.Binding
+	Paste                   key.Binding
+}
+
+// DefaultKeyMap is the default set of key bindings for navigating and acting
+// upon the textarea.
```
Comment (maaslalani) [inline textinput/textinput.go:59]:
```suggestion
// upon the textinput.
```

File: textinput/textinput.go [hunk 2]
```diff
@@ -37,6 +38,41 @@ const (
 // ValidateFunc is a function that returns an error if the input is invalid.
 type ValidateFunc func(string) error
 
+// KeyMap is the key bindings for different actions within the textarea.
```
Comment (maaslalani) [inline textinput/textinput.go:41]:
```suggestion
// KeyMap is the key bindings for different actions within the textinput.
```

File: textinput/textinput.go [hunk 3]
```diff
@@ -71,6 +107,9 @@ type Model struct {
 	// viewport. If 0 or less this setting is ignored.
 	Width int
 
+	// KeyMap encodes the keybindings recognized by the widget.
```
Comment (maaslalani) [inline textinput/textinput.go:110]:
```suggestion
	// KeyMap encodes the keybindings recognized by the textinput.
```

File: textinput/textinput.go [hunk 4]
```diff
@@ -473,68 +513,49 @@ func (m Model) Update(msg tea.Msg) (Model, tea.Cmd) {
...
 				}
 			}
-		case tea.KeyLeft, tea.KeyCtrlB:
-			if msg.Alt { // alt+left arrow, back one word
-				m.wordLeft()
-				break
-			}
-			if m.pos > 0 { // left arrow, ^F, back one character
+		case key.Matches(msg, m.KeyMap.WordBackward):
+			m.wordLeft()
```
Comment (maaslalani) [inline textinput/textinput.go:529]: I think we should rename this method to match the others.

```suggestion
			m.wordBackward()
```

File: textinput/textinput.go [hunk 5]
```diff
@@ -473,68 +513,49 @@ func (m Model) Update(msg tea.Msg) (Model, tea.Cmd) {
...
 				m.SetCursor(m.pos - 1)
 			}
-		case tea.KeyRight, tea.KeyCtrlF:
-			if msg.Alt { // alt+right arrow, forward one word
-				m.wordRight()
-				break
-			}
-			if m.pos < len(m.value) { // right arrow, ^F, forward one character
+		case key.Matches(msg, m.KeyMap.WordForward):
+			m.wordRight()
```
Comment (maaslalani) [inline textinput/textinput.go:535]: We should rename this method to match the others.

```suggestion
			m.wordForward()
```

**Review by maaslalani — APPROVED** [review]

**Comment by maaslalani** [comment]
Just some small notes but the PR works perfectly, really great work @knz. Thank you so much! Really nice to see that `Alt` check basically disappear from the default case.

File: textinput/textinput.go (hunk 1 above)
Comment (knz) [reply textinput/textinput.go]: done

File: textinput/textinput.go (hunk 2 above)
Comment (knz) [reply textinput/textinput.go]: done

File: textinput/textinput.go (hunk 3 above)
Comment (knz) [reply textinput/textinput.go:110]: done

File: textinput/textinput.go (hunk 4 above)
Comment (knz) [reply textinput/textinput.go]: done

File: textinput/textinput.go (hunk 5 above)
Comment (knz) [reply textinput/textinput.go]: done

**Pull request merged**